---
title: "GeoMx Analysis"
output: html_document
date: "2023-10-27"
---

```{r setup, include=FALSE}
library(NanoStringNCTools)
library(GeomxTools)
library(GeoMxWorkflows)
library(knitr)
library(dplyr)
library(ggforce)
library(readxl)
library(reshape2)
library(tidyverse)
library(pheatmap)
library(lmerTest)
library(reshape2)
library(cowplot)

pseudocount <- 1

log_adj <- function(x){
  return(log2(x + pseudocount))
}

calc_CV <- function(x) {sd(x) / mean(x)}
```

## Setup

```{r get_files, echo=FALSE}
## get DCC, PKC, annotation file
DCCFiles <- dir(file.path("data/"), pattern = ".dcc$",
                full.names = TRUE, recursive = TRUE)
PKCFiles <- "src/Hs_R_NGS_WTA_v1.0.pkc"
SampleAnnotationFile <- read_excel("data/Annotation_GeoMx.xlsx", sheet = "Annotation_GeoMx")

meta_lookup <- read_excel("data/BPC-19-SK-0668-HTMA-438-map.xlsx", sheet = "Lookup", col_names=TRUE) %>%
  distinct(`DONOR BLOCK ID`, .keep_all=TRUE)

clinical_anno <- readxl::read_xlsx('tables/BCBM TMA Report 2023-12-31_final.xlsx', sheet='Patient')

meta_map <- read_excel("data/BPC-19-SK-0668-HTMA-438-map.xlsx", sheet = "BPC-19-SK-0668 HTMA 438_B", col_names=TRUE) %>%
  pivot_longer(cols = c('X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8'), names_to = "Xval", values_to = "Sample_Name") %>%
  mutate(Slide_Number = paste(Xval, Y, sep=""))

SampleAnnotationFile <- SampleAnnotationFile %>%
  mutate(Slide_Number = str_sub(Roi, 6, -2L)) %>%
  left_join(select(meta_map, Slide_Number, Sample_Name), by = "Slide_Number") %>%
  left_join(meta_lookup, by = c("Sample_Name"="DONOR BLOCK ID")) %>%
  left_join(clinical_anno, by = c("Slide_Number"="roi"))

# write.csv(SampleAnnotationFile, 'output/annotation-file-metadata.csv', row.names = FALSE)

SampleAnnotationFilePath <- "output/annotation-file-metadata.xlsx"

```

```{r create_object}
demoData <-
    readNanoStringGeoMxSet(dccFiles = DCCFiles,
                           pkcFiles = PKCFiles,
                           phenoDataFile = SampleAnnotationFilePath,
                           phenoDataSheet = "Annotation_GeoMx",
                           phenoDataDccColName = "Sample_ID",
                           protocolDataColNames = c("Roi", "Aoi"),
                           experimentDataColNames = c("Panel"))

# DSP-10016600005648-A-C02.dcc has no counts!
```

```{r check_prelim}
pkcs <- annotation(demoData)
modules <- gsub(".pkc", "", pkcs)
data.frame(PKCs = pkcs, modules = modules)

pData(demoData)
dplyr::count(pData(demoData), Comment, SITE)
```

```{r QC_flags}
QC_params <-
    list(minSegmentReads = 1000, # Minimum number of reads (1000)
         percentTrimmed = 80,    # Minimum % of reads trimmed (80%)
         percentStitched = 80,   # Minimum % of reads stitched (80%)
         percentAligned = 75,    # Minimum % of reads aligned (80%)
         percentSaturation = 50, # Minimum sequencing saturation (50%)
         minNegativeCount = 1,   # Minimum negative control counts (10)
         maxNTCCount = 9000,     # Maximum counts in NTC well (1000)
         minNuclei = 20,         # Minimum # of nuclei estimated (100)
         minArea = 1000)         # Minimum segment area (5000)

demoData <-
    setSegmentQCFlags(demoData, 
                      qcCutoffs = QC_params)

QCResults <- protocolData(demoData)[["QCFlags"]] %>%
  select(-LowNegatives)
# QCResults <- subset(QCResults, select = -c("LowNegatives"))

flag_columns <- colnames(QCResults)
flag_columns <- flag_columns[!flag_columns == 'LowNegatives']

QC_Summary <- data.frame(Pass = colSums(!QCResults[, flag_columns]),
                         Warning = colSums(QCResults[, flag_columns]))

QCResults$QCStatus <- apply(QCResults, 1L, function(x) {
    ifelse(sum(x) == 0L, "PASS", "WARNING")
})

QC_Summary["TOTAL FLAGS", ] <-
    c(sum(QCResults[, "QCStatus"] == "PASS"),
      sum(QCResults[, "QCStatus"] == "WARNING"))

QC_Summary
```

```{r Segment_QC}
col_by <- "Comment"

# Graphical summaries of QC statistics plot function
QC_histogram <- function(assay_data = NULL,
                         annotation = NULL,
                         fill_by = NULL,
                         thr = NULL,
                         scale_trans = NULL) {
    plt <- ggplot(assay_data,
                  aes_string(x = paste0("unlist(`", annotation, "`)"),
                             fill = fill_by)) +
        geom_histogram(bins = 50) +
        geom_vline(xintercept = thr, lty = "dashed", color = "black") +
        theme_bw() + guides(fill = "none") +
        facet_wrap(as.formula(paste("~", fill_by)), nrow = 4) +
        labs(x = annotation, y = "Segments, #", title = annotation)
    if(!is.null(scale_trans)) {
        plt <- plt +
            scale_x_continuous(trans = scale_trans)
    }
    plt
}

QC_histogram(sData(demoData), "Trimmed (%)", col_by, QC_params$percentTrimmed)
# ggsave("figs/geomx-trimmed.png")
QC_histogram(sData(demoData), "Stitched (%)", col_by, QC_params$percentStitched)
# ggsave("figs/geomx-stitched.png")
QC_histogram(sData(demoData), "Aligned (%)", col_by, QC_params$percentAligned)
# ggsave("figs/geomx-aligned.png")
QC_histogram(sData(demoData), "Saturated (%)", col_by, QC_params$percentSaturation)
# ggsave("figs/geomx-saturated.png")
QC_histogram(sData(demoData), "Area", col_by, QC_params$minArea, scale_trans = "log10")
# ggsave("figs/geomx-area.png")
QC_histogram(sData(demoData), "Nuclei", col_by, QC_params$minNuclei)
# ggsave("figs/geomx-nuclei.png")

negativeGeoMeans <- 
    esBy(negativeControlSubset(demoData), 
         GROUP = "Module", 
         FUN = function(x) { 
             assayDataApply(x, MARGIN = 2, FUN = ngeoMean, elt = "exprs") 
         }) 
protocolData(demoData)[["NegGeoMean"]] <- negativeGeoMeans

negCols <- paste0("NegGeoMean_", modules)
pData(demoData)[, negCols] <- sData(demoData)[["NegGeoMean"]]
for(ann in negCols) {
    plt <- QC_histogram(pData(demoData), ann, col_by, QC_params$minNegativeCount, scale_trans = "log10")
    print(plt)
}

dim(demoData) 
```

```{r filtered_samples}
# remove flagged samples
pData(demoData) <- pData(demoData) %>% mutate(Flag = QCResults$QCStatus)

demoData <- demoData[, !is.na(pData(demoData)$Sample_Name)]

dplyr::count(pData(demoData), Flag, Comment, SITE)
dim(demoData[, pData(demoData)$Flag == "PASS"])
dplyr::count(subset(pData(demoData), Flag == "PASS"), SAMPLE_ID)
```

```{r Probe_QC}
demoData <- setBioProbeQCFlags(demoData, 
                               qcCutoffs = list(minProbeRatio = 0.1,
                                                percentFailGrubbs = 20), 
                               removeLocalOutliers = TRUE)

ProbeQCResults <- fData(demoData)[["QCFlags"]]
qc_df <- data.frame(Passed = sum(rowSums(ProbeQCResults[, -1]) == 0),
                    Global = sum(ProbeQCResults$GlobalGrubbsOutlier),
                    Local = sum(rowSums(ProbeQCResults[, -2:-1]) > 0
                                & !ProbeQCResults$GlobalGrubbsOutlier))

ProbeQCPassed <- 
    subset(demoData, 
           fData(demoData)[["QCFlags"]][,c("LowProbeRatio")] == FALSE &
               fData(demoData)[["QCFlags"]][,c("GlobalGrubbsOutlier")] == FALSE)

demoData <- ProbeQCPassed

demoData <- aggregateCounts(demoData)
dim(demoData)

target_demoData <- demoData[, pData(demoData)$Flag == "PASS"]
# write.csv(exprs(target_demoData), "output/geomx-count-matrix.csv")
```

## Analysis of All AOI
  
```{r All_LOQ}
cutoff <- 2
minLOQ <- 2

# Calculate LOQ per module tested
LOQ <- data.frame(row.names = colnames(target_demoData))
for(module in modules) {
    vars <- paste0(c("NegGeoMean_", "NegGeoSD_"),
                   module)
    if(all(vars[1:2] %in% colnames(pData(target_demoData)))) {
        LOQ[, module] <-
            pmax(minLOQ,
                 pData(target_demoData)[, vars[1]] * 
                     pData(target_demoData)[, vars[2]] ^ cutoff)
    }
}
pData(target_demoData)$LOQ <- LOQ

LOQ_Mat <- c()
for(module in modules) {
    ind <- fData(target_demoData)$Module == module
    Mat_i <- t(esApply(target_demoData[ind, ], MARGIN = 1,
                       FUN = function(x) {
                           x > LOQ[, module]
                       }))
    LOQ_Mat <- rbind(LOQ_Mat, Mat_i)
}
# ensure ordering since this is stored outside of the geomxSet
LOQ_Mat <- LOQ_Mat[fData(target_demoData)$TargetName, ]
```

```{r All_segment_filtering}
# Save detection rate information to pheno data
pData(target_demoData)$GenesDetected <- 
    colSums(LOQ_Mat, na.rm = TRUE)
pData(target_demoData)$GeneDetectionRate <-
    pData(target_demoData)$GenesDetected / nrow(target_demoData)

# Determine detection thresholds: 1%, 5%, 10%, 15%, >15%
pData(target_demoData)$DetectionThreshold <- 
    cut(pData(target_demoData)$GeneDetectionRate,
        breaks = c(0, 0.01, 0.05, 0.1, 0.15, 1),
        labels = c("<1%", "1-5%", "5-10%", "10-15%", ">15%"))

# stacked bar plot of different cut points (1%, 5%, 10%, 15%)
ggplot(pData(target_demoData),
       aes(x = DetectionThreshold)) +
    geom_bar(aes(fill = Comment)) +
    geom_text(stat = "count", aes(label = ..count..), vjust = -0.5) +
    theme_bw() +
    scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
    labs(x = "Gene Detection Rate",
         y = "Segments, #",
         fill = "Segment Type")

# ggsave("figs/all-segment-detection-rate-by-region.png")

dplyr::count(pData(target_demoData), Comment, DetectionThreshold)
```

```{r All_gene_filtering}
tdd <- target_demoData[, pData(target_demoData)$GeneDetectionRate >= 0.1 & pData(target_demoData)$Flag == "PASS"]

# Calculate detection rate:
LOQ_Mat <- LOQ_Mat[, colnames(tdd)]
fData(tdd)$DetectedSegments <- rowSums(LOQ_Mat, na.rm = TRUE)
fData(tdd)$DetectionRate <-
    fData(tdd)$DetectedSegments / nrow(pData(tdd))

plot_detect <- data.frame(Freq = c(1, 5, 10, 20, 30, 50))
plot_detect$Number <-
    unlist(lapply(c(0.01, 0.05, 0.1, 0.2, 0.3, 0.5),
                  function(x) {sum(fData(tdd)$DetectionRate >= x)}))
plot_detect$Rate <- plot_detect$Number / nrow(fData(tdd))
rownames(plot_detect) <- plot_detect$Freq

ggplot(plot_detect, aes(x = as.factor(Freq), y = Rate, fill = Rate)) +
    geom_bar(stat = "identity") +
    geom_text(aes(label = formatC(Number, format = "d", big.mark = ",")),
              vjust = 1.6, color = "black", size = 4) +
    scale_fill_gradient2(low = "orange2", mid = "lightblue",
                         high = "dodgerblue3", midpoint = 0.65,
                         limits = c(0,1),
                         labels = scales::percent) +
    theme_bw() +
    scale_y_continuous(labels = scales::percent, limits = c(0,1),
                       expand = expansion(mult = c(0, 0))) +
    labs(x = "% of Segments",
         y = "Genes Detected, % of Panel > LOQ")

# ggsave("figs/strict-all-percent-segments-detected.png")

negativeProbefData <- subset(fData(tdd), CodeClass == "Negative")
neg_probes <- unique(negativeProbefData$TargetName)

target_demoData <- 
    tdd[fData(tdd)$DetectionRate >= 0.2 |
                        fData(tdd)$TargetName %in% neg_probes, ]

dim(target_demoData)
dplyr::count(pData(target_demoData), Comment, SITE)
```

```{r All_Q3_norm}
# Graph Q3 value vs negGeoMean of Negatives
ann_of_interest <- "SITE"
Stat_data <- 
    data.frame(row.names = colnames(exprs(target_demoData)),
               Segment = colnames(exprs(target_demoData)),
               Annotation = pData(target_demoData)[, ann_of_interest],
               Q3 = unlist(apply(exprs(target_demoData), 2,
                                 quantile, 0.75, na.rm = TRUE)),
               NegProbe = exprs(target_demoData)[neg_probes, ])
Stat_data_m <- melt(Stat_data, measure.vars = c("Q3", "NegProbe"),
                    variable.name = "Statistic", value.name = "Value")

plt1 <- ggplot(Stat_data_m,
               aes(x = Value, fill = Statistic)) +
    geom_histogram(bins = 40) + theme_bw() +
    scale_x_continuous(trans = "log2") +
    facet_wrap(~Annotation, nrow = 1) + 
    scale_fill_brewer(palette = 3, type = "qual") +
    labs(x = "Counts", y = "Segments, #")

plt2 <- ggplot(Stat_data,
               aes(x = NegProbe, y = Q3, color = Annotation)) +
    geom_abline(intercept = 0, slope = 1, lty = "dashed", color = "darkgray") +
    geom_point() + guides(color = "none") + theme_bw() +
    scale_x_continuous(trans = "log2") + 
    scale_y_continuous(trans = "log2") +
    theme(aspect.ratio = 1) +
    labs(x = "Negative Probe GeoMean, Counts", y = "Q3 Value, Counts")

plt3 <- ggplot(Stat_data,
               aes(x = NegProbe, y = Q3 / NegProbe, color = Annotation)) +
    geom_hline(yintercept = 1, lty = "dashed", color = "darkgray") +
    geom_point() + theme_bw() +
    scale_x_continuous(trans = "log2") + 
    scale_y_continuous(trans = "log2") +
    theme(aspect.ratio = 1) +
    labs(x = "Negative Probe GeoMean, Counts", y = "Q3/NegProbe Value, Counts")

btm_row <- plot_grid(plt2, plt3, nrow = 1, labels = c("B", ""),
                     rel_widths = c(0.43,0.57))
plot_grid(plt1, btm_row, ncol = 1, labels = c("A", ""))


target_demoData <- normalize(target_demoData,
                             norm_method = "quant", 
                             desiredQuantile = .75,
                             toElt = "q_norm")

assayDataElement(object = target_demoData, elt = "log_q") <-
    assayDataApply(target_demoData, 2, FUN = log_adj, elt = "q_norm")

# png(file="figs/boxplot-raw-counts-all-aoi.png", width=600, height=300)
boxplot(assayDataElement(target_demoData, elt="exprs"),
        col = "#9EDAE5", main = "Raw Counts",
        log="y", ylim = c(0.001, max(exprs(target_demoData))),
        ylab = "Counts", outline=FALSE, xaxt="n")
axis(tick=FALSE, side=1)
# dev.off()

# png(file="figs/boxplot-q3-norm-all-aoi.png", width=600, height=300)
boxplot(assayDataElement(target_demoData, elt="q_norm"),
        col = "#2CA02C", main = "Q3 Normalized Counts",
        log="y", ylim = c(0.001, max(assayDataElement(target_demoData, elt="q_norm"))),
        ylab = "Counts", outline=FALSE, xaxt="n")
axis(tick=FALSE, side=1)
# dev.off()

dim(pData(target_demoData))
dplyr::count(pData(target_demoData), Comment, SITE)
```


```{r PCA_filtered_genes}
pcamat <- assayDataElement(target_demoData, elt="log_q")

pca_cv <- apply(pcamat, 1, calc_CV)
pca_cv <- sort(pca_cv, decreasing = TRUE)
pca_goi <- names(pca_cv)[1:1000]

pc_out <- prcomp(t(pcamat[pca_goi,]), scale.=FALSE, center=TRUE)

pData(target_demoData)[, c("PC1", "PC2")] <- pc_out$x[,c("PC1", "PC2")]

temp <- pData(target_demoData)
sampleids <- temp$SAMPLE_ID
sampleids <- gsub("_BR", "", sampleids)
brs <- sampleids %in% c("BCBM_1", "BCBM_22", "BCBM_27")
sampleids[!brs] <- "None"
temp$MATCH <- sampleids

dplyr::count(temp, SITE, MATCH)
             
ggplot(temp,
       aes(x = PC1, y = PC2, color=SITE)) +
    geom_point(size = 2) +
    theme_bw() +
    xlab(paste("PC1 (", 100*summary(pc_out)[["importance"]][2,1], "%)", sep="")) +
    ylab(paste("PC2 (", 100*summary(pc_out)[["importance"]][2,2], "%)", sep="")) # +
    # scale_color_manual(values=c(None='grey', BCBM_1='purple', BCBM_22='green', BCBM_27='red'))

# ggsave("figs/strict-all-pca-by-site.png")

scree <- summary(pc_out)[["importance"]][2:3,]
scree <- as.data.frame(t(scree))
scree <- scree %>%
  rownames_to_column(var="PC") %>%
  mutate(PC = factor(PC, levels=PC)) %>%
  rename(Proportion = `Proportion of Variance`, Cumulative = `Cumulative Proportion`)

ggplot(scree[1:10,]) + geom_col(aes(x=PC, y=Proportion)) +
geom_line(aes(x=PC, y=Cumulative), color='red', size=1, group=1) +
      geom_point(aes(x=PC, y=Cumulative), color='red', size=3) + theme_bw()

# ggsave("figs/strict-all-scree-plot.png")

save.image(file = "all_qc.RData")
```